### ISSUE TITLE
[Proposal] Standardise JSON Responses

### ISSUE THREAD

**Issue by DanielDarrenJones** [laravel/internals#603]
# The Problem:  
Across Laravel there is not currently a consistant JSON reponse format, this makes interacting with the different JSON responses in Laravel more complex than it should be.

Laravel is quickly allowing for greater decoupling between it's frontend and backend. While the majority of websites still use session authentication, a growing number websites are moving to interface with the applications via their public API directly through the likes of Passport.

# Examples: 

## Model Response
``` json
{
  "id": 1,
  "name": "Miracle Wisoky",
  "email": "maggio.lucienne@example.com",
  "created_at": "2017-05-12 17:48:23",
  "updated_at": "2017-05-12 17:48:23"
}
```

## Collection Response
``` json
[
  {
    "id": 1,
    "name": "Miracle Wisoky",
    "email": "maggio.lucienne@example.com",
    "created_at": "2017-05-12 17:48:23",
    "updated_at": "2017-05-12 17:48:23"
  },
  {
    "id": 2,
    "name": "Alexandro Abshire",
    "email": "lon.hegmann@example.org",
    "created_at": "2017-05-12 17:49:40",
    "updated_at": "2017-05-12 17:49:40"
  },
  {
    "id": 3,
    "name": "Martine Leffler",
    "email": "nicole60@example.com",
    "created_at": "2017-05-12 17:51:37",
    "updated_at": "2017-05-12 17:51:37"
  },
  {
    "id": 4,
    "name": "Jaunita Johnson",
    "email": "lorenzo93@example.org",
    "created_at": "2017-05-12 17:55:03",
    "updated_at": "2017-05-12 17:55:03"
  }
]
```

## Paginated Collection Response
``` json
{
  "users": {
    "total": 4,
    "per_page": 5,
    "current_page": 1,
    "last_page": 1,
    "next_page_url": null,
    "prev_page_url": null,
    "from": 1,
    "to": 4,
    "data": [
      {
        "id": 1,
        "name": "Miracle Wisoky",
        "email": "maggio.lucienne@example.com",
        "created_at": "2017-05-12 17:48:23",
        "updated_at": "2017-05-12 17:48:23"
      },
      {
        "id": 2,
        "name": "Alexandro Abshire",
        "email": "lon.hegmann@example.org",
        "created_at": "2017-05-12 17:49:40",
        "updated_at": "2017-05-12 17:49:40"
      },
      {
        "id": 3,
        "name": "Martine Leffler",
        "email": "nicole60@example.com",
        "created_at": "2017-05-12 17:51:37",
        "updated_at": "2017-05-12 17:51:37"
      },
      {
        "id": 4,
        "name": "Jaunita Johnson",
        "email": "lorenzo93@example.org",
        "created_at": "2017-05-12 17:55:03",
        "updated_at": "2017-05-12 17:55:03"
      }
    ]
  }
}
```

## Validation Error Response
``` json
{
  "user_id": [
    "The user id field is required."
  ],
  "name": [
    "The name field is required."
  ],
  "email": [
    "The email field is required."
  ]
}
```

## Passport Error Response
``` json
{
  "error": "invalid_credentials",
  "message": "The user credentials were incorrect."
}
```

# The Solution  
I propose as a community we settle on an agreed best practice API format, which the laravel framework will adopt as it's standard JSON reponse type, in doing so it will make both building, and consuming API's easier for websites which are built on Laravel.

Laravel is in a great spot to push a best practice standard, and help shape the use of clean, easy to use API's across the web, due to it's large userbase and strong community.

In order for this initiative to be truly effective there will need to be two major changes:
1. The unified JSON format will need to be decided and implemented across the laravel framework.
2. A system for transforming data to the JSON format will need to be implemented to allow users to both quickly and easily create their JSON responses.

I would appreciate feedback from anyone with extensive experiance developing API's and I welcome discussion below, I look forward to hearing from you all!

**Comment by scrwdnet** [comment]
Something like [jsonapi.org](http://jsonapi.org/) would be a good standard to follow, this way we don't have to come up with our own specification and can concentrate on the implementation.

Should people be allowed to decide if they want a more complex spec or to keep the current simple spec? Personally if I was writing a mini 'internal' API then I'd like to keep it simple.

**Comment by barryvdh** [comment]
What about http://fractal.thephpleague.com/ ?

**Comment by scrwdnet** [comment]
@barryvdh yeah their [JSON API Serializer](http://fractal.thephpleague.com/serializers/#jsonapiserializer) works with the jsonapi.org spec. Are you suggesting Laravel implements this or how a developer should be doing it?

**Comment by barryvdh** [comment]
I'm suggesting that there is a perfectly fine 3rd party package to handle this, so not sure if Laravel needs to handle this.

**Comment by scrwdnet** [comment]
@barryvdh I guess so but @DanielDarrenJones is suggesting Laravel adopts a spec to normalise it's responses. It might be better for Laravel to implement then everyone has access to a spec based response? ¯\_(ツ)_/¯

Why did car manufactures implement bluetooth connectivity into their media centres when you can get a perfectly fine 3rd party bluetooth connector? 😛

**Comment by DanielDarrenJones** [comment]
In my opinion I think Laravel should have a fractal wrapper, much like how it wraps up other components such as stripe in cashier, it can use this across the board to create its internal responses and could have a system to override the return schema if people really want to go against it.

I like the idea of JSON api, but it feels rather heavy I think a slightly lighter version of it would be good for Laravel if thats possible?

**Comment by adiachenko** [comment]
@scrwdnet Fractal doesn't work with JSON API spec, it only looks like it does. Try ember-data or any other tool that was actually designed to consume data from JSON API compliant back-end.

Anyway, the spec itself is overcomplicated with its relative linkeage, bizzare content negotiation and other requirements that go way beyond the scope of standardising the sctructure of json responses. It reminds me of SOAP and other such monstrosities.

**Comment by salomoni** [comment]
Could this be implemented in the middleware layer?

Would be easier to swap between example just between API versions.

**Comment by themsaid** [comment]
@DanielDarrenJones Take a look at https://github.com/laravel/framework/pull/19449

**Comment by DanielDarrenJones** [comment]
@themsaid that looks absolutely perfect, very simple and easy to use.

Awesome work Mohamed! Thanks :)

Would be good to add a scaffold to artisan make to build these up and generate a folder to store them?

**Comment by themsaid** [comment]
Yup upcoming next when/if this was merged :)

**Comment by DCzajkowski** [comment]
I think this can be closed as of [Laravel 5.5 release](https://twitter.com/christophrumpel/status/902551807121125376).
